### PR TITLE
Promote CowData to 64 bits

### LINKED ISSUE #62585: Godot crash in ResourceSaver after 1073741824 bytes
### Godot version

4.0.alpha c41e4b1

### System information

Windows 10, 64-bit

### Issue description

I ran into this when saving a large scn with lots of textures, and used "Make Local" so all textures got embedded.

I tried saving this scene, and it gave an error in resize
```
ERROR: Condition "p_size < 0" is true. Returning: ERR_INVALID_PARAMETER
   at: CowData<unsigned char>::resize (S:\repo\godot-fire\core/templates/cowdata.h:261)
```
After this error, it proceeded to crash.
```
 	godot.windows.opt.tools.64.exe!FileAccessCompressed::store_8(unsigned char p_dest) Line 339	C++
>	godot.windows.opt.tools.64.exe!FileAccess::store_buffer(const unsigned char * p_src, unsigned __int64 p_length) Line 552	C++
 	godot.windows.opt.tools.64.exe!ResourceFormatSaverBinaryInstance::write_variant(Ref<FileAccess> f, const Variant & p_property, HashMap<Ref<Resource>,int,HashMapHasherDefault,HashMapComparatorDefault<Ref<Resource> >,DefaultTypedAllocator<HashMapElement<Ref<Resource>,int> > > & resource_map, HashMap<Ref<Resource>,int,HashMapHasherDefault,HashMapComparatorDefault<Ref<Resource> >,DefaultTypedAllocator<HashMapElement<Ref<Resource>,int> > > & external_resources, HashMap<StringName,int,HashMapHasherDefault,HashMapComparatorDefault<StringName>,DefaultTypedAllocator<HashMapElement<StringName,int> > > & string_map, const PropertyInfo & p_hint) Line 1657	C++
 	godot.windows.opt.tools.64.exe!ResourceFormatSaverBinaryInstance::write_variant(Ref<FileAccess> f, const Variant & p_property, HashMap<Ref<Resource>,int,HashMapHasherDefault,HashMapComparatorDefault<Ref<Resource> >,DefaultTypedAllocator<HashMapElement<Ref<Resource>,int> > > & resource_map, HashMap<Ref<Resource>,int,HashMapHasherDefault,HashMapComparatorDefault<Ref<Resource> >,DefaultTypedAllocator<HashMapElement<Ref<Resource>,int> > > & external_resources, HashMap<StringName,int,HashMapHasherDefault,HashMapComparatorDefault<StringName>,DefaultTypedAllocator<HashMapElement<StringName,int> > > & string_map, const PropertyInfo & p_hint) Line 1637	C++
 	godot.windows.opt.tools.64.exe!ResourceFormatSaverBinaryInstance::save(const String & p_path, const Ref<Resource> & p_resource, unsigned int p_flags) Line 2081	C++
 	godot.windows.opt.tools.64.exe!ResourceFormatSaverBinary::save(const String & p_path, const Ref<Resource> & p_resource, unsigned int p_flags) Line 2106	C++
 	godot.windows.opt.tools.64.exe!ResourceSaver::save(const String & p_path, const Ref<Resource> & p_resource, unsigned int p_flags) Line 110	C++
```

CowData resizes to the next power of two, which is 2147483648, and overflows a signed int.

I think everything here should be made less sloppy. Array operations should always use unsigned, not signed values. Integers should be 64-bit where possible: GDScript already uses 64-bit, so why are they downgraded to 32-bit in C++? Finally, FileAccess should check the return value of `resize()` and fail gracefully.

### Steps to reproduce

1. Import a very heavy scene from glTF, such that it has a few GB of assets.
2. Use "Make local" or do a compessed export using a script.

Sorry I don't have more specific steps. The asset I used is not redistributable, and the save was automated by a script in order to enable ZSTD compression.

### Minimal reproduction project

N/A

### LINKED ISSUE #77256: Error when resizing PackedByteArray bigger than int32 limit
### Godot version

4.0.3

### System information

Windows 11 64-Bit

### Issue description

The PackedByteArray can't be bigger than 2147483647. Following [this](https://github.com/godotengine/godot/issues/46842#issuecomment-842216412) issue it should be fixed in Godot 4.0 RC 1, but running the code below in Godot 4.0.3 results in this error.

```
E 0:00:00:0834   Node.gd:9 @ _ready(): Condition "p_size < 0" is true. Returning: ERR_INVALID_PARAMETER
  <C++-Quelle>   ./core/templates/cowdata.h:262 @ resize()
  <Stacktrace>   Node.gd:9 @ _ready()

```

### Steps to reproduce

Run this code:
```
	var INT32_MAX = 2147483647 # 2^31 - 1
	var byte_arr = PackedByteArray()
	byte_arr.resize(INT32_MAX) # OK
	byte_arr.resize(INT32_MAX + 1) # error
```

### Minimal reproduction project

[PackedByteArrayError.zip](https://github.com/godotengine/godot/files/11520507/PackedByteArrayError.zip)

### LINKED ISSUE #80533: Attempting to read large file as string causes segfault
### Godot version

4.1.1.stable, master [4714e9589]

### System information

Godot v4.2.dev (4714e9589) - Arch Linux #1 ZEN SMP PREEMPT_DYNAMIC Tue, 08 Aug 2023 22:13:48 +0000 - Wayland - Vulkan (Forward+) - dedicated AMD Radeon RX 5700 (RADV NAVI10) () - AMD Ryzen 7 5800X3D 8-Core Processor (16 Threads)

### Issue description

I attempted to read a large JSON file (1GB+) using `FileAccess.get_file_as_string`, but the result is a crash. Here's a backtrace from a debug build [4714e9589] and some info, not sure what to make of this

```
Loading...

Thread 1 "godot.linuxbsd." received signal SIGSEGV, Segmentation fault.
String::parse_utf8 (this=0x7fffffffa5f8, 
    p_utf8=0x7fff2ffff020 "[{\"p\":[-0.019984539598226547,0.3165023624897003,0.6493667960166931],\"c\":[0.15294925029263085,0.1603185939029318,0.11855615533237046],\"s\":[-5.519049167633057,-5.86762809753418,-4.488621234893799],\"q\":["..., p_len=1093019370, p_skip_cr=false) at core/string/ustring.cpp:1880
1880            dst[str_size] = 0;
(gdb) bt
#0  String::parse_utf8 (this=0x7fffffffa5f8, 
    p_utf8=0x7fff2ffff020 "[{\"p\":[-0.019984539598226547,0.3165023624897003,0.6493667960166931],\"c\":[0.15294925029263085,0.1603185939029318,0.11855615533237046],\"s\":[-5.519049167633057,-5.86762809753418,-4.488621234893799],\"q\":["..., p_len=1093019370, p_skip_cr=false) at core/string/ustring.cpp:1880
#1  0x000055555cae1c27 in FileAccess::get_file_as_string (p_path=..., r_error=0x0) at core/io/file_access.cpp:770
#2  0x000055555cae631e in FileAccess::_get_file_as_string (p_path=...) at core/io/file_access.h:227
#3  0x000055555caf36ae in call_with_variant_args_static_ret<String, String const&, 0ul> (p_method=0x55555cae62e7 <FileAccess::_get_file_as_string(String const&)>, p_args=0x7fffffffa720, r_ret=..., r_error=...)
    at ./core/variant/binder_common.h:766
#4  0x000055555caf0cbd in call_with_variant_args_static_ret_dv<String, String const&> (p_method=0x55555cae62e7 <FileAccess::_get_file_as_string(String const&)>, p_args=0x7fffffffa838, p_argcount=1, r_ret=..., r_error=..., 
    default_values=...) at ./core/variant/binder_common.h:989
#5  0x000055555caeec5a in MethodBindTRS<String, String const&>::call (this=0x5555616f8240, p_object=0x0, p_args=0x7fffffffa838, p_arg_count=1, r_error=...) at ./core/object/method_bind.h:719
#6  0x000055555870c3db in GDScriptFunction::call (this=0x555563c0aa10, p_instance=0x555563bfaba0, p_args=0x0, p_argcount=0, r_err=..., p_state=0x0) at modules/gdscript/gdscript_vm.cpp:1934
#7  0x000055555859a338 in GDScriptInstance::callp (this=0x555563bfaba0, p_method=..., p_args=0x0, p_argcount=0, r_error=...) at modules/gdscript/gdscript.cpp:1892
#8  0x000055555aaeec0c in Node::_gdvirtual__ready_call<false> (this=0x555563bf7930) at scene/main/node.h:322
#9  0x000055555aac7d65 in Node::_notification (this=0x555563bf7930, p_notification=13) at scene/main/node.cpp:186
#10 0x00005555582b77d4 in Node::_notificationv (this=0x555563bf7930, p_notification=13, p_reversed=false) at ./scene/main/node.h:49
#11 0x00005555582d7147 in Node3D::_notificationv (this=0x555563bf7930, p_notification=13, p_reversed=false) at ./scene/3d/node_3d.h:52
#12 0x000055555d092445 in Object::notification (this=0x555563bf7930, p_notification=13, p_reversed=false) at core/object/object.cpp:798
#13 0x000055555aac823e in Node::_propagate_ready (this=0x555563bf7930) at scene/main/node.cpp:230
#14 0x000055555aac81be in Node::_propagate_ready (this=0x555563b84780) at scene/main/node.cpp:221
#15 0x000055555aadae3e in Node::_set_tree (this=0x555563b84780, p_tree=0x555563b84280) at scene/main/node.cpp:2909
#16 0x000055555ab18704 in SceneTree::initialize (this=0x555563b84280) at scene/main/scene_tree.cpp:449
#17 0x0000555558035c2d in OS_LinuxBSD::run (this=0x7fffffffdbf0) at platform/linuxbsd/os_linuxbsd.cpp:900
#18 0x000055555802c274 in main (argc=3, argv=0x7fffffffe1b8) at platform/linuxbsd/godot_linuxbsd.cpp:74
(gdb) p dst
$1 = 0x7fff27ffe020 U""
(gdb) p str_size
$2 = 1093019370
(gdb) p dst[str_size]
Cannot access memory at address 0x80002c987bc8
(gdb) p _cowdata.size()
$3 = 1093019371
```

After some manual binary searching 0x7ffff7fff000 seems to pop up...
```
(gdb) p dst[872416247]
$35 = 0 U'\000'
(gdb) p dst[872416248]
Cannot access memory at address 0x7ffff7fff000
```

### Steps to reproduce

```
extends Node3D

# Called when the node enters the scene tree for the first time.
func _ready():
	print("Loading...")
	var data = FileAccess.get_file_as_string("/path/to/out.json")
	print("Loaded")
	print(len(data))


# Called every frame. 'delta' is the elapsed time since the previous frame.
func _process(delta):
	pass
```

In case the specific file is needed it's available [here](https://april.sapples.net/out.json) (warning: it's big)

### Minimal reproduction project

-

### PR DESCRIPTION
**Pull request by reduz**
Fixes a lot of bugs, please help me fill the list.
```
< I welcome the 64 bits overlords >
 ---------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||
```

After this is merged, other areas of the engine can be fixed, such as Image, which also uses 32 bits offsets and causes crashes in some reasonable scenarios (such as baking lightmaps), but it depends on this one. Serialization also may need to be fixed.

*Bugsquad edit:*
- Fixes #62585
- Fixes #77256
- Fixes #80533

### CODE REVIEW

**Comment by Mickeon** [comment]
Would fix https://github.com/godotengine/godot/issues/45296 and https://github.com/godotengine/godot/issues/54679 for sure.

**Comment by AThousandShips** [comment]
Note that this requires `_get_alloc_size` to be updated unless we still treat this as having a capped size in the 32 bit range, as it only handles 32 bits, `next_power_of_2` needs to be updated

Also needs a godot-cpp companion

File: core/templates/cowdata.h [hunk 1]
```diff
@@ -200,43 +202,44 @@ void CowData<T>::_unref(void *p_data) {
...
 
-	uint32_t rc = refc->get();
+	uint64_t rc = refc->get();
 	if (unlikely(rc > 1)) {
 		/* in use by more than me */
-		uint32_t current_size = *_get_size();
+		uint64_t current_size = *_get_size();
 
-		uint32_t *mem_new = (uint32_t *)Memory::alloc_static(_get_alloc_size(current_size), true);
+		uint64_t *mem_new = (uint64_t *)Memory::alloc_static(_get_alloc_size(current_size) + ALLOC_PAD, false);
```
Comment (RandomShaper) [inline core/templates/cowdata.h:239]: `Memory::alloc_static()` adds a 16-byte pre-padding, of which the last 8 bytes are availale for "users". When `CowData` was 32-bit, those 8 bytes were enough to store the two header `uint32_t`s. With this PR, `CowData` is adding 16 more bytes, which gives a total of 24 extra bytes. Assuming the ref count has to become 64-bit as well, we're wasting 8 bytes.

Possible options:
- a) Redefine `PAD_ALIGN` in `memory.h` to `24` and remove the newly added `ALLOC_PAD` in `CowData`.
- b) Redefine `PAD_ALIGN` in `memory.h` to `8` (assumming the extra bytes aren't used elsewhere) and let `CowData` deal with all the pre-padding (i.e., keep the current `ALLOC_PAD` of `16`).

File: core/templates/cowdata.h [hunk 2]
```diff
@@ -65,24 +65,26 @@ class CowData {
...
+	_FORCE_INLINE_ uint64_t *_get_size() const {
 		if (!_ptr) {
 			return nullptr;
 		}
 
-		return reinterpret_cast<uint32_t *>(_ptr) - 1;
+		return reinterpret_cast<uint64_t *>(_ptr) - 1;
 	}
 
 	_FORCE_INLINE_ size_t _get_alloc_size(size_t p_elements) const {
```
Comment (AThousandShips) [inline core/templates/cowdata.h:90]: I'd suggest switching to `uint64_t` here as `size_t` isn't guaranteed to be 64 bits though it usually is, better stay safe IMO

**Review by RandomShaper — COMMENTED** [review]
Looks good overall.

I'm wondering if this is neglecting 32-bit builds (they don't need or won't support 64-bit stuff such as > 3 GiB allocations and maybe not even 64-bit atomics) and so `size_t` should be used intead.

**UPDATE:** Per @AThousandShips, `size_t` won't necessarilly be 64-bit on 64-bit targets, so maybe something else that matches `sizeof(void *)`, but you get the idea.

File: core/templates/cowdata.h (hunk 1 above)
Comment (AThousandShips) [reply core/templates/cowdata.h:239]: Only on debug builds, it doesn't add the padding otherwise, as the second parameter is false

Comment (AThousandShips) [reply core/templates/cowdata.h:239]: Note that the 8 bytes are used already, forget where but I've seen it, will look

Comment (reduz) [reply core/templates/cowdata.h:239]: Yes I was going to mention this, it only adds 16 bytes on debug builds, but not on release, where it remains 16.

Comment (RandomShaper) [reply core/templates/cowdata.h:239]: I missed the argument has been changed to `false`. So, yeah, that's OK. Let's just find out (as a separate endeavor) if the extra pre-padding from `memory.h` is still needed somewhere.

Comment (AThousandShips) [reply core/templates/cowdata.h:239]: It is, in `memnew_arr` 🙂

File: core/templates/cowdata.h (hunk 2 above)
Comment (reduz) [reply core/templates/cowdata.h]: I ended up using a typedef for the size, as suggested.

**Comment by reduz** [comment]
I changed to use a typedef for the size, depending on compiler, so it remains 32 bits on 32 bit platforms.

**Comment by reduz** [comment]
@AThousandShips regarding GodotCPP, I am not certain if something has to change on that side.

**Comment by AThousandShips** [comment]
It does, it redeclares the type, unless we're okay with lack of parity over there

File: core/templates/vector.h [hunk 3]
```diff
@@ -157,40 +158,40 @@ class Vector {
 		return ret;
 	}
 
-	Vector<T> slice(int p_begin, int p_end = INT_MAX) const {
+	Vector<T> slice(Size p_begin, Size p_end = INT64_MAX) const {
```
Comment (AThousandShips) [inline core/templates/vector.h:161]: Need a macro for the max as well, I'd suggest a `SIZE_MAX/MIN` and `USIZE_MAX/MIN` though might clash with definitions existing already

**Comment by reduz** [comment]
@AThousandShips ah I completely forgot we strive to use the same syntax in GodotCPP than in Godot modules.. Will have to ask @bruvzg to lend a hand changing the code there once this one is merged.

So I guess this PR is "kind of" compat breaking.. though it likely only affects GodotCPP, not the rest of the extensions.

File: core/typedefs.h [hunk 4]
```diff
@@ -139,9 +139,22 @@ inline void __swap_tmpl(T &x, T &y) {
...
+/* GCC, Clang */
+#if defined(__i386__) || defined(__ILP32__) || (defined(__SIZEOF_POINTER__) && (__SIZEOF_POINTER__ == 4))
+#define IS_32_BIT
+#endif
+#elif defined(_MSC_VER)
+/* Microsoft Visual C++ */
+#if defined(_M_IX86)
+#define IS_32_BIT
+#endif
+#endif
```
Comment (akien-mga) [inline core/typedefs.h:152]: Sounds like a pretty old stackoverflow answer :)

Needs thorough review to do it right. I doubt we care about `__ILP32__`, but we do care about 32-bit ARM.

CC @bruvzg

File: core/typedefs.h [hunk 5]
```diff
@@ -139,9 +139,22 @@ inline void __swap_tmpl(T &x, T &y) {
...
+
 /* Functions to handle powers of 2 and shifting. */
 
 // Function to find the next power of 2 to an integer.
+
```
Comment (akien-mga) [inline core/typedefs.h:157]: Why? :) The comment directly qualifies the method under it.

File: core/typedefs.h (hunk 4 above)
Comment (reduz) [reply core/typedefs.h:152]: I checked it and it should be alright, basically anything GCC/LLVM uses __SIZEOF_POINTER__ anyway, so we are safe (and this also contemplates emscripten).

File: core/templates/cowdata.h [hunk 6]
```diff
@@ -64,45 +64,74 @@ class CowData {
 	template <class TV, class VV>
 	friend class VMap;
 
+public:
+#ifdef IS_32_BIT
+	typedef int32_t Size;
+	typedef uint32_t USize;
+#else
+	typedef int64_t Size;
+	typedef uint64_t USize;
```
Comment (akien-mga) [inline core/templates/cowdata.h:73]: Might need more specific names, I'm worried about conflicts with potential Windows or macOS global defines.

Comment (reduz) [reply core/templates/cowdata.h:73]: It's in the local class context so it should be fine as long as it passes CI.

File: core/typedefs.h (hunk 5 above)
Comment (reduz) [reply core/typedefs.h:157]: oops, sorry

File: core/templates/cowdata.h (hunk 6 above)
Comment (bruvzg) [reply core/templates/cowdata.h:73]: Is this extra complexity really necessary. 32-bit systems won't benefit for it, but using 64-bit ints should work fine on 32-bit systems (as long as it's not cast to `void*` directly).

**Comment by TokisanGames** [comment]
Thank God. I haven't tested yet, but will link up related issues. Lack of 64-bit resources is a problem for Terrain3D users with big terrains.

So hopefully this will fix our issue, crashing when saving large terrains in a >1gb resource file. 
https://github.com/TokisanGames/Terrain3D/issues/159

And Godot issue https://github.com/godotengine/godot/issues/62585

Possibility superseding or related to PRs 
https://github.com/godotengine/godot/pull/74582
https://github.com/godotengine/godot/pull/77306

File: core/typedefs.h (hunk 4 above)
Comment (bruvzg) [reply core/typedefs.h:152]: 32-bit ARM is still supported for Android (not sure if we do official builds, but the option is still there) as well as for Windows (completely useless in practice).

File: core/templates/cowdata.h (hunk 6 above)
Comment (bruvzg) [reply core/templates/cowdata.h:73]: And if we want to do it this way for some reason, why not use `size_t` and `ptrdiff_t` directly?

File: core/typedefs.h (hunk 4 above)
Comment (Calinou) [reply core/typedefs.h:152]: WebAssembly is also 32-bit as of writing :slightly_smiling_face:

Comment (AThousandShips) [reply core/typedefs.h:152]: The CI will tell in fact!

Comment (RandomShaper) [reply core/typedefs.h:152]: Couldn't this be solved with some compile-time magic? Some SFINAE, or a more modern approach, involving pointer size and type aliasing.

Comment (reduz) [reply core/typedefs.h:152]: As I said, anything other than Windows is either GCC or LLVM which uses __SIZEOF_POINTER__, so this is already solved problem.

File: core/templates/cowdata.h (hunk 6 above)
Comment (reduz) [reply core/templates/cowdata.h:73]: I don't mind leaving it 64 bits even for 32 bits systems.

File: core/templates/cowdata.h [hunk 7]
```diff
@@ -64,45 +64,74 @@ class CowData {
...
+		}
+
+		--x;
+		x |= x >> 1;
+		x |= x >> 2;
+		x |= x >> 4;
+		x |= x >> 8;
+		x |= x >> 16;
+		if (sizeof(USize) == 8) {
+			x |= x >> 32;
```
Comment (AThousandShips) [inline core/templates/cowdata.h:89]: This leaves the maximum size as `2^33`, is this intended as a limit to not go too big or an oversight?

**Comment by AThousandShips** [comment]
While changing the size values of `String` might be risky because it's exposed (granted so is `Vector`, indirectly), `VMap` and `VSet` uses `Vector` internally, should we unify all data structures using these sizes at least the non-exposed ones?

**Comment by AThousandShips** [comment]
As long as the iteration cases aren't broken it should be safe enough

File: core/templates/cowdata.h [hunk 8]
```diff
@@ -64,45 +64,71 @@ class CowData {
...
+		}
+
+		return ++x;
+	}
+
+	static constexpr USize ALLOC_PAD = 16; // Always 16 bye alignment.
```
Comment (AThousandShips) [inline core/templates/cowdata.h:92]:
```suggestion
	static constexpr USize ALLOC_PAD = 16; // Always 16 byte alignment.
```

**Review by AThousandShips — APPROVED** [review]
The code looks good to me, the sizing checks should work correctly with the new po2 method and my previous fix for the resize should still handle that just with a new range

Needs to ensure that any iteration being done isn't broken, might need a fix of any `for (int...)` loops using `Vector` so we don't get broken data etc., as part of the fixes involved in this are for cases where data does run out

Otherwise looks good, if everything else works right

**Comment by reduz** [comment]
This is now ready to merge. 

It is to note that this PR will break GodotCPP GDextension compatibility, yet again.

This is an unfortunate consequence of the fact that GodotCPP accesses internals it should technically not for the sake of being compatible with modules. We should maybe consider having a different set of bindings, or do some things in GodotCPP different to avoid this scenario.

**Comment by akien-mga** [comment]
There's a MSVC warning to fix:
```
Error: .\core/templates/ring_buffer.h(200): error C2220: the following warning is treated as an error
Warning: .\core/templates/ring_buffer.h(200): warning C4334: '<<': result of 32-bit shift implicitly converted to 64 bits (was 64-bit shift intended?)
```

---

And the "Test project converter" in the C# build fails with a C# stacktrace:
https://github.com/godotengine/godot/actions/runs/7472530025/job/20334943529?pr=86730

But I don't know if it's actually due to a C# issue, or just a Godot error that would happen on a non-dotnet build with just a different error (that project converter test is only run on the "Editor w/ Mono" Linux build).

The test is just this, so this can be attempted locally to see if the error is reproducible:
```
mkdir converter_test
cd converter_test
touch project.godot
godot --headless --validate-conversion-3to4
cd ..
rm converter_test -rf
```

**Comment by adamscott** [comment]
cc. @Bromeon @dsnopek

**Comment by AThousandShips** [comment]
Would suggest running with `-k` on windows if you can to catch all the cases as there's likely many

**Comment by reduz** [comment]
I haven't used MSVC to compile in an eternity, will check during the week.

**Comment by dsnopek** [comment]
> It is to note that this PR will break GodotCPP GDextension compatibility, yet again.
>
> This is an unfortunate consequence of the fact that GodotCPP accesses internals it should technically not for the sake of being compatible with modules. We should maybe consider having a different set of bindings, or do some things in GodotCPP different to avoid this scenario.

I don't think this is exactly true.

It will only break godot-cpp, if godot-cpp is interacting with a `String` or `Packed*Array` or similar, that has more items than a 32-bit index can access. It isn't actually accessing any internals that it shouldn't, it just uses implementations of `CowData` and `Vector` etc that are derived from the code for the same classes in Godot itself, so they suffer from the same limitations.

I think with bruvzg's godot-cpp PR https://github.com/godotengine/godot-cpp/pull/1357, we should be good going forward. And I think we can cherrypick it for the 4.2 and 4.1 versions of godot-cpp.

But if a GDExtension compiled with an older godot-cpp is used with a newer Godot, AND it's passed a `String` or `Packed*Array` with too many items, then there could be some breakage. However, this feels like a small enough edge case to probably be acceptable?

**Comment by raulsntos** [comment]
To the extent that this changes the memory layout of Variant types, it has an impact on C# interop structs.

In C# we define interop structs that must match the memory layout of the C++ types so they can be reinterpreted and consumed from C#. For example, we get the size of a Godot String like this:

https://github.com/godotengine/godot/blob/107f2961ccfac179af7682eb5f6e7ea91e80040c/modules/mono/glue/GodotSharp/GodotSharp/Core/NativeInterop/InteropStructs.cs#L477

This is basically copied from the C++ code (`CowData::get_size`):

https://github.com/godotengine/godot/blob/107f2961ccfac179af7682eb5f6e7ea91e80040c/core/templates/cowdata.h#L85

But this has changed in this PR since now the size is 64-bit, so it needs to be updated in C# too.

As long as the changes are kept internally, it should not break compatibility but changing the user-facing API would be a problem. And if any of these types actually has a size that doesn't fit in a 32-bit `int` then it will break C#[^1].

<details><summary>Here's a somewhat naive diff that should fix the stackoverflow you are getting on CI:</summary>

```diff
diff --git a/modules/mono/glue/GodotSharp/GodotSharp/Core/NativeInterop/InteropStructs.cs b/modules/mono/glue/GodotSharp/GodotSharp/Core/NativeInterop/InteropStructs.cs
index d5d9404ed1b..c806263edb5 100644
--- a/modules/mono/glue/GodotSharp/GodotSharp/Core/NativeInterop/InteropStructs.cs
+++ b/modules/mono/glue/GodotSharp/GodotSharp/Core/NativeInterop/InteropStructs.cs
@@ -474,7 +474,7 @@ namespace Godot.NativeInterop
         public readonly unsafe int Size
         {
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            get => _ptr != IntPtr.Zero ? *((int*)_ptr - 1) : 0;
+            get => _ptr != IntPtr.Zero ? (int)(*((ulong*)_ptr - 1)) : 0;
         }
     }
 
@@ -725,7 +725,7 @@ namespace Godot.NativeInterop
             public readonly unsafe int Size
             {
                 [MethodImpl(MethodImplOptions.AggressiveInlining)]
-                get => _ptr != null ? *((int*)_ptr - 1) : 0;
+                get => _ptr != null ? (int)(*((ulong*)_ptr - 1)) : 0;
             }
         }
 
@@ -875,7 +875,7 @@ namespace Godot.NativeInterop
         public readonly unsafe int Size
         {
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            get => _ptr != null ? *((int*)_ptr - 1) : 0;
+            get => _ptr != null ? (int)(*((ulong*)_ptr - 1)) : 0;
         }
     }
 
@@ -939,7 +939,7 @@ namespace Godot.NativeInterop
         public readonly unsafe int Size
         {
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            get => _ptr != null ? *((int*)_ptr - 1) : 0;
+            get => _ptr != null ? (int)(*((ulong*)_ptr - 1)) : 0;
         }
     }
 
@@ -971,7 +971,7 @@ namespace Godot.NativeInterop
         public readonly unsafe int Size
         {
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            get => _ptr != null ? *((int*)_ptr - 1) : 0;
+            get => _ptr != null ? (int)(*((ulong*)_ptr - 1)) : 0;
         }
     }
 
@@ -1003,7 +1003,7 @@ namespace Godot.NativeInterop
         public readonly unsafe int Size
         {
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            get => _ptr != null ? *((int*)_ptr - 1) : 0;
+            get => _ptr != null ? (int)(*((ulong*)_ptr - 1)) : 0;
         }
     }
 
@@ -1035,7 +1035,7 @@ namespace Godot.NativeInterop
         public readonly unsafe int Size
         {
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            get => _ptr != null ? *((int*)_ptr - 1) : 0;
+            get => _ptr != null ? (int)(*((ulong*)_ptr - 1)) : 0;
         }
     }
 
@@ -1067,7 +1067,7 @@ namespace Godot.NativeInterop
         public readonly unsafe int Size
         {
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            get => _ptr != null ? *((int*)_ptr - 1) : 0;
+            get => _ptr != null ? (int)(*((ulong*)_ptr - 1)) : 0;
         }
     }
 
@@ -1099,7 +1099,7 @@ namespace Godot.NativeInterop
         public readonly unsafe int Size
         {
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            get => _ptr != null ? *((int*)_ptr - 1) : 0;
+            get => _ptr != null ? (int)(*((ulong*)_ptr - 1)) : 0;
         }
     }
 
@@ -1131,7 +1131,7 @@ namespace Godot.NativeInterop
         public readonly unsafe int Size
         {
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            get => _ptr != null ? *((int*)_ptr - 1) : 0;
+            get => _ptr != null ? (int)(*((ulong*)_ptr - 1)) : 0;
         }
     }
 

```

</details>

[^1]: More context on this: https://github.com/dotnet/runtime/issues/12221

**Comment by reduz** [comment]
@raulsntos Yeah I guess the Mono bindings should not really be accessing these internals. We have explicit functions exposed in the extension API now to avoid these situations, so I suppose it would be just a matter to move to that eventually (when Mono is moved to extension)

**Comment by raulsntos** [comment]
I don't want to get too off-topic, but I don't agree that the move to GDExtensions would help here. The API exposed to retrieve this data would have a bigger cost than accessing the internals directly as we are doing today so we wouldn't change that in the new system.

And to be honest that's not the problem, we just need to update the C# internals to match the C++ internals (similar to how godot-cpp is doing). I shared a diff that does just that and doesn't break compatibility since it doesn't change publicly exposed API.

The main issue is that C# types like `System.String` or arrays define their length as 32-bit integers so they can't possible work with data that is bigger than what can be represented with a `int32_t`[^1].

To put it in simple terms, if we have a `String` coming from Godot that has more items than what can fit in a 32-bit length then we wouldn't be able to marshal that into a C# `String`.

[^1]: https://github.com/dotnet/runtime/issues/12221

**Comment by dsnopek** [comment]
> And to be honest that's not the problem, we just need to update the C# internals to match the C++ internals (similar to how godot-cpp is doing).

Just to be clear (although, it probably isn't necessary :-)) this isn't what the godot-cpp change is doing. godot-cpp _doesn't_ directly access Godot's internals, it always goes through the GDExtension interface.

The change to godot-cpp is doing two things:

1. Updating the C++ classes that exist entirely on the godot-cpp side to account for the possibility that the data on the Godot side could have a large enough number of elements that it requires a 64-bit integer to count/index them
2. Updating the equivalent `CowData` and `Vector` in godot-cpp to match Godot, so that you can use them in your GDExtension exactly the same as you would in Godot itself, removing the 32-bit limitations of the previous implementation. This doesn't affect interop with Godot, this is for where these templates are used purely on the godot-cpp side.

**Comment by akien-mga** [comment]
> Fixes a lot of bugs, please help me fill the list.

Trying to review some of the ones mentioned in this PR, linked issues, or some I could find searching for references to CowData and 32/64 bits.

- #45296

`3.x` issue so it won't be solved by this (and likely won't ever be solved).

- #48399

Reproducible in `master`. This PR doesn't fix it because BitMap now has guards against using a size bigger than `INT32_MAX`. This PR should however make it possible to refactor BitMap to use `int64_t`.

- #51154

Not fixed by this PR. The stacktrace changes from `CowData<_AtlasWorkRectResult>::get(int)` to `CowData<_AtlasWorkRectResult>::get(long)` but it's still crashing. Likely another one that _could_ be fixed now as a follow-up.

- #54679

Seems related indeed, but I tested and this PR does _not_ solve the issue. It still crashes in the same place.
Relevant Image code is still using `int` for sizes so I assume it needs further work after this PR to actually fix the crash.
```
[3] Image::initialize_data(int, int, bool, Image::Format) (/home/akien/Projects/godot/godot.git/./core/io/image.cpp:2187 (discriminator 4))
```

- #62585

Tested with the MRP from https://github.com/godotengine/godot/issues/62585#issuecomment-1450900291, this fixes it.
I don't get a crash however before this PR in a 2 weeks old `masterbuild, but I do get a `p_size < 0` error which is now fixed.

- #77256

Fixed by this PR.

- #80533

Fixed by this PR.

**Comment by akien-mga** [comment]
Diff that should make the PR pass CI (two more MSVC warning fixes, and @raulsntos' C# changes):
https://github.com/akien-mga/godot/commit/6eff4f655343d2e0a917f75eba1dcafee713cfd0

**Review by akien-mga — APPROVED** [review]
Tested locally, seems to work well!

As I pointed out, further work will be needed at least for Image and BitMap to make their size handling 64-bit too. Possibly other classes based on CowData internally.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**